File: code/client/core/source/operation-graph/GenerateResultManager.cpp
// <copyright file="GenerateResultManager.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <memory>
#include <set>
#include <stdexcept>

export module Soup.Core:GenerateResultManager;

import Opal;
import :FileSystemState;
import :GenerateResult;
import :GenerateResultReader;
import :GenerateResultWriter;

using namespace Opal;

namespace Soup::Core
{
	/// <summary>
	/// The generate result state manager
	/// </summary>
	export class GenerateResultManager
	{
	public:
		/// <summary>
		/// Load the generate result state from the provided directory
		/// </summary>
		static bool TryLoadState(
			const Path& generateResultFile,
			GenerateResult& result,
			FileSystemState& fileSystemState)
		{
			// Open the file to read from
			std::shared_ptr<System::IInputFile> file;
			if (!System::IFileSystem::Current().TryOpenRead(generateResultFile, true, file))
			{
				Log::Info("Generate result file does not exist");
				return false;
			}

			// Read the contents of the build state file
			try
			{
				result = GenerateResultReader::Deserialize(file->GetInStream(), fileSystemState);
				return true;
			}
			catch(std::runtime_error& ex)
			{
				Log::Error(ex.what());
				return false;
			}
			catch(...)
			{
				Log::Error("Failed to parse generate result");
				return false;
			}
		}

		/// <summary>
		/// Save the generate result state for the provided directory
		/// </summary>
		static void SaveState(
			const Path& generateResultFile,
			GenerateResult& state,
			const FileSystemState& fileSystemState)
		{
			auto targetFolder = generateResultFile.GetParent();

			// Update the generate result referenced files
			auto files = std::set<FileId>();
			for (auto& operationReference : state.GetEvaluateGraph().GetOperations())
			{
				auto& operation = operationReference.second;
				files.insert(operation.DeclaredInput.begin(), operation.DeclaredInput.end());
				files.insert(operation.DeclaredOutput.begin(), operation.DeclaredOutput.end());
				files.insert(operation.ReadAccess.begin(), operation.ReadAccess.end());
				files.insert(operation.WriteAccess.begin(), operation.WriteAccess.end());
			}

			for (auto& operationProxyReference : state.GetOperationProxies())
			{
				auto& operationProxy = operationProxyReference.second;
				files.insert(operationProxy.DeclaredInput.begin(), operationProxy.DeclaredInput.end());
				files.insert(operationProxy.ReadAccess.begin(), operationProxy.ReadAccess.end());
			}

			// Open the file to write to
			auto file = System::IFileSystem::Current().OpenWrite(generateResultFile, true);

			// Write the build state to the file stream
			GenerateResultWriter::Serialize(state, files, fileSystemState, file->GetOutStream());
		}
	};
}

File: code/generate-sharp/utilities/operation-graph/GenerateResultManager.cs
// <copyright file="GenerateResultManager.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using Opal;
using Opal.System;
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Text;
using Path = Opal.Path;

namespace Soup.Build.Utilities;

/// <summary>
/// The generate result manager
/// </summary>
public static class GenerateResultManager
{
	/// <summary>
	/// Load the generate result from the provided directory
	/// </summary>
	public static bool TryLoadState(
		Path GenerateResultFile,
		FileSystemState fileSystemState,
		[MaybeNullWhen(false)] out GenerateResult result)
	{
		// Verify the requested file exists
		if (!LifetimeManager.Get<IFileSystem>().Exists(GenerateResultFile))
		{
			Log.Info("Generate result file does not exist");
			result = null;
			return false;
		}

		// Open the file to read from
		using var file = LifetimeManager.Get<IFileSystem>().OpenRead(GenerateResultFile);
		using var reader = new System.IO.BinaryReader(file.GetInStream(), Encoding.UTF8, true);

		// Read the contents of the build state file
		try
		{
			var loadedResult = GenerateResultReader.Deserialize(reader);

			// Map up the incoming file ids to the active file system state ids
			var activeFileIdMap = new Dictionary<FileId, FileId>();
			for (var i = 0; i < loadedResult.ReferencedFiles.Count; i++)
			{
				var fileReference = loadedResult.ReferencedFiles[i];
				var activeFileId = fileSystemState.ToFileId(fileReference.Path);
				activeFileIdMap.Add(fileReference.FileId, activeFileId);

				// Update the referenced id
				fileReference.FileId = activeFileId;
			}

			// Update all of the operations
			foreach (var operationReference in loadedResult.EvaluateGraph.Operations)
			{
				var operation = operationReference.Value;
				UpdateFileIds(operation.DeclaredInput, activeFileIdMap);
				UpdateFileIds(operation.DeclaredOutput, activeFileIdMap);
			}

			foreach (var operationReference in loadedResult.OperationProxies)
			{
				var operation = operationReference.Value;
				UpdateFileIds(operation.DeclaredInput, activeFileIdMap);
				operation.ResultFile = UpdateFileId(operation.ResultFile, activeFileIdMap);
			}

			result = loadedResult;
			return true;
		}
		catch (InvalidOperationException)
		{
			Log.Error("Failed to parse operation graph");
			result = null;
			return false;
		}
	}

	/// <summary>
	/// Save the operation state for the provided directory
	/// </summary>
	public static void SaveState(
		Path GenerateResultFile,
		GenerateResult state,
		FileSystemState fileSystemState)
	{
		// Update the operation graph referenced files
		var files = new HashSet<FileId>();
		foreach (var operationReference in state.EvaluateGraph.Operations)
		{
			var operation = operationReference.Value;
			files.UnionWith(operation.DeclaredInput);
			files.UnionWith(operation.DeclaredOutput);
			files.UnionWith(operation.ReadAccess);
			files.UnionWith(operation.WriteAccess);
		}

		foreach (var operationProxyReference in state.OperationProxies)
		{
			var operationProxy = operationProxyReference.Value;
			files.UnionWith(operationProxy.DeclaredInput);
			_ = files.Add(operationProxy.ResultFile);
			files.UnionWith(operationProxy.ReadAccess);
		}

		var referencedFiles = new List<(FileId FileId, Path Path)>();
		foreach (var fileId in files)
		{
			referencedFiles.Add((fileId, fileSystemState.GetFilePath(fileId)));
		}

		state.ReferencedFiles.Clear();
		state.ReferencedFiles.AddRange(referencedFiles);

		// Open the file to write to
		using var fileStream = System.IO.File.Open(
			GenerateResultFile.ToString(),
			System.IO.FileMode.Create,
			System.IO.FileAccess.Write);
		using var writer = new System.IO.BinaryWriter(fileStream);

		// Write the build state to the file stream
		GenerateResultWriter.Serialize(state, writer);
	}

	private static void UpdateFileIds(IList<FileId> fileIds, Dictionary<FileId, FileId> activeFileIdMap)
	{
		for (var i = 0; i < fileIds.Count; i++)
		{
			var findActiveFileId = activeFileIdMap[fileIds[i]];
			fileIds[i] = findActiveFileId;
		}
	}

	private static FileId UpdateFileId(FileId fileId, Dictionary<FileId, FileId> activeFileIdMap)
	{
		var findActiveFileId = activeFileIdMap[fileId];
		return findActiveFileId;
	}
}